### DISCUSSION TITLE
Dealing with external error types that are not `Send`

### DISCUSSION THREAD

**Discussion by ilyvion** [zkat/miette#105] [Q&A] [answered]
Hi.

Couldn't find any information on this particular problem, so I'm hoping someone here can shed some light on it. I'm trying to use `miette`'s `into_diagnostic()` with an `std::sync::RwLock`'s `Result`s, and it's yelling at me. 😅

The issue is that `read()` and `write()` calls on `RwLock` returns `Result<Guard, PoisonError<Guard>>` where `Guard` is either `RwLockReadGuard<'_, T>` or `RwLockWriteGuard<'_, T>`, which are both explicitly `!Send` (and even if that wasn't an issue, they're not `'static` either), which is incompatible with `IntoDiagnostic`'s rather strict `E: Error + Send + Sync + 'static` bound.

I solved it by doing `.map_err(|e| miette!("{}", e))` (for some definitions of "solve"), but I was wondering if there's a less "throw miette's fancy error rendering out the window" way of doing this?

It's not exactly required, but here's the what Rust tells me is wrong for the sake of completeness:
<details>
  <summary><code>cargo check</code> output</summary>
 
``` 
error[E0599]: the method `into_diagnostic` exists for enum `Result<RwLockReadGuard<'_, Arc<Configuration>>, PoisonError<RwLockReadGuard<'_, Arc<Configuration>>>>`, but its trait bounds were not satisfied
   --> src\configuration.rs:29:57
    |
29  |     let configuration_read_guard = configuration.read().into_diagnostic()?;
    |                                                         ^^^^^^^^^^^^^^^ method cannot be called on `Result<RwLockReadGuard<'_, Arc<Configuration>>, PoisonError<RwLockReadGuard<'_, Arc<Configuration>>>>` due to unsatisfied trait bounds
    |
   ::: C:\Users\alexs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\std\src\sync\poison.rs:83:1
    |
83  | pub struct PoisonError<T> {
    | ------------------------- doesn't satisfy `_: Send`
    |
   ::: C:\Users\alexs\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\core\src\result.rs:503:1
    |
503 | pub enum Result<T, E> {
    | --------------------- doesn't satisfy `_: IntoDiagnostic<RwLockReadGuard<'_, Arc<Configuration>>, PoisonError<RwLockReadGuard<'_, Arc<Configuration>>>>`
    |
    = note: the following trait bounds were not satisfied:
            `PoisonError<RwLockReadGuard<'_, Arc<Configuration>>>: Send`
            which is required by `Result<RwLockReadGuard<'_, Arc<Configuration>>, PoisonError<RwLockReadGuard<'_, Arc<Configuration>>>>: IntoDiagnostic<RwLockReadGuard<'_, Arc<Configuration>>, PoisonError<RwLockReadGuard<'_, Arc<Configuration>>>>`
```
</details>

**Comment by zkat** [comment, marked as answer]
Unfortunately, I have to give the same answer here that `eyre` gives: https://github.com/yaahc/eyre/issues/51

That is, you can work around this by making a wrapper type, which is not ideal, but should work?

**Reply by ilyvion** [reply]
That might've worked if it wasn't also non-`'static`. But I think just doing `.map_err(|e| miette!("{}", e))` will be sufficient, actually, because that error type isn't very fancy when it comes to it. All it does in its `display()` is `"poisoned lock: another task failed inside".fmt(f)`, so I don't think I'll be losing much by doing so.

Thanks for getting back to me. 😊
